examples/vision/edsr.py

```diff
@@ -0,0 +1,276 @@
+[EDSR Paper](https://arxiv.org/abs/1707.02921)
+"""
+
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow.keras.optimizers.schedules import PiecewiseConstantDecay
+
+AUTOTUNE = tf.data.AUTOTUNE
+
+import warnings
+
+warnings.simplefilter("ignore")
+
+import matplotlib.pyplot as plt
+from PIL import Image
+import numpy as np
+import cv2
+import os
+
+"""
+# Data Agumentation
+"""
+
+# Fliping left and right
+def flip_left_right(lr_img, hr_img):
+    rn = tf.random.uniform(shape=(), maxval=1)
+    return tf.cond(
+        rn < 0.5,
+        lambda: (lr_img, hr_img),
+        lambda: (tf.image.flip_left_right(lr_img), tf.image.flip_left_right(hr_img)),
+    )
+
+
+# Rotating both images by 90 degree
+def random_rotate(lr_img, hr_img):
+    rn = tf.random.uniform(shape=(), maxval=4, dtype=tf.int32)
+    return tf.image.rot90(lr_img, rn), tf.image.rot90(hr_img, rn)
+
+
+# Cropping high resolution image to 96x96 and low resolution to 24x24
+def random_crop(lr_img, hr_img, hr_crop_size=96, scale=4):
+    lr_crop_size = hr_crop_size // scale
+    lr_img_shape = tf.shape(lr_img)[:2]
+
+    lr_width = tf.random.uniform(
+        shape=(), maxval=lr_img_shape[1] - lr_crop_size + 1, dtype=tf.int32
+    )
+    lr_height = tf.random.uniform(
+        shape=(), maxval=lr_img_shape[0] - lr_crop_size + 1, dtype=tf.int32
+    )
+
+    hr_width = lr_width * scale
+    hr_height = lr_height * scale
+
+    lr_img_cropped = lr_img[
+        lr_height : lr_height + lr_crop_size, lr_width : lr_width + lr_crop_size
+    ]
+    hr_img_cropped = hr_img[
+        hr_height : hr_height + hr_crop_size, hr_width : hr_width + hr_crop_size
+    ]
+
+    return lr_img_cropped, hr_img_cropped
+
+
+"""
+# Loading Data using tensorflow datasets
+"""
+
+div2k_data = tfds.image.Div2k(config="bicubic_x4")
+div2k_data.download_and_prepare()
+
+train = div2k_data.as_dataset(split="train", as_supervised=True)
+train_cache = train.cache()
+
+ds = train_cache
+
+ds = ds.map(lambda lr, hr: random_crop(lr, hr, scale=4), num_parallel_calls=AUTOTUNE)
+ds = ds.map(random_rotate, num_parallel_calls=AUTOTUNE)
+ds = ds.map(flip_left_right, num_parallel_calls=AUTOTUNE)
+
+ds = ds.batch(16)
+ds = ds.repeat(None)
+ds = ds.prefetch(buffer_size=AUTOTUNE)
+
+lr, hr = next(iter(ds))
+
+"""
+# High Resolution Images
+"""
+
+plt.figure(figsize=(10, 10))
+for i in range(9):
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(hr[i].numpy().astype("uint8"))
+    plt.title(hr[i].shape)
+    plt.axis("off")
+
+"""
+# Low Resolution Images
+"""
+
+plt.figure(figsize=(10, 10))
+for i in range(9):
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(lr[i].numpy().astype("uint8"))
+    plt.title(lr[i].shape)
+    plt.axis("off")
+
+"""
+# Normalization and usefull functions
+As, the paper mentioned we will subtract image by mean rgb of div2k dataset
+"""
+
+DIV2K_RGB_MEAN = np.array([0.4488, 0.4371, 0.4040]) * 255
```

Comment:
Mean subtraction is not helpful compared to plain scaling. Let's just use scaling.